File: lib/src/internal/routes/magic_link.dart
import '/src/sdk/routes.dart';
import '/src/types/others.dart';
import '/src/types/responses.dart';
import '/src/internal/http/descope_client.dart';
import 'shared.dart';

class MagicLink implements DescopeMagicLink {
  final DescopeClient client;

  MagicLink(this.client);

  @override
  Future<String> signUp({required DeliveryMethod method, required String loginId, SignUpDetails? details, String? redirectUrl}) async {
    return (await client.magicLinkSignUp(method, loginId, details, redirectUrl)).convert(method);
  }

  @override
  Future<String> signIn({required DeliveryMethod method, required String loginId, String? redirectUrl, SignInOptions? options}) async {
    return (await client.magicLinkSignIn(method, loginId, redirectUrl, options)).convert(method);
  }

  @override
  Future<String> signUpOrIn({required DeliveryMethod method, required String loginId, String? redirectUrl, SignInOptions? options}) async {
    return (await client.magicLinkSignUpOrIn(method, loginId, redirectUrl, options)).convert(method);
  }

  @override
  Future<String> updateEmail({required String email, required String loginId, String? redirectUrl, required String refreshJwt, UpdateOptions? options}) async {
    return (await client.magicLinkUpdateEmail(email, loginId, redirectUrl, refreshJwt, options)).convert(DeliveryMethod.email);
  }

  @override
  Future<String> updatePhone({required String phone, required DeliveryMethod method, required String loginId, String? redirectUrl, required String refreshJwt, UpdateOptions? options}) async {
    return (await client.magicLinkUpdatePhone(phone, method, loginId, redirectUrl, refreshJwt, options)).convert(method);
  }

  @override
  Future<AuthenticationResponse> verify({required String token}) async {
    return (await client.magicLinkVerify(token)).convert();
  }
}

File: ios/Classes/descope-swift-sdk/internal/routes/MagicLink.swift

import Foundation

final class MagicLink: DescopeMagicLink {
    let client: DescopeClient
    
    init(client: DescopeClient) {
        self.client = client
    }
    
    func signUp(with method: DeliveryMethod, loginId: String, details: SignUpDetails?, redirectURL: String?) async throws(DescopeError) -> String {
        return try await client.magicLinkSignUp(with: method, loginId: loginId, details: details, redirectURL: redirectURL).convert(method: method)
    }
    
    func signIn(with method: DeliveryMethod, loginId: String, redirectURL: String?, options: [SignInOptions]) async throws(DescopeError) -> String {
        let (refreshJwt, loginOptions) = try options.convert()
        return try await client.magicLinkSignIn(with: method, loginId: loginId, redirectURL: redirectURL, refreshJwt: refreshJwt, options: loginOptions).convert(method: method)
    }
    
    func signUpOrIn(with method: DeliveryMethod, loginId: String, redirectURL: String?, options: [SignInOptions]) async throws(DescopeError) -> String {
        let (refreshJwt, loginOptions) = try options.convert()
        return try await client.magicLinkSignUpOrIn(with: method, loginId: loginId, redirectURL: redirectURL, refreshJwt: refreshJwt, options: loginOptions).convert(method: method)
    }
    
    func updateEmail(_ email: String, loginId: String, redirectURL: String?, refreshJwt: String, options: UpdateOptions) async throws(DescopeError) -> String {
        return try await client.magicLinkUpdateEmail(email, loginId: loginId, redirectURL: redirectURL, refreshJwt: refreshJwt, options: options).convert(method: .email)
    }
    
    func updatePhone(_ phone: String, with method: DeliveryMethod, loginId: String, redirectURL: String?, refreshJwt: String, options: UpdateOptions) async throws(DescopeError) -> String {
        return try await client.magicLinkUpdatePhone(phone, with: method, loginId: loginId, redirectURL: redirectURL, refreshJwt: refreshJwt, options: options).convert(method: method)
    }
    
    func verify(token: String) async throws(DescopeError) -> AuthenticationResponse {
        return try await client.magicLinkVerify(token: token).convert()
    }
}
